shunting-yard-replicator/src/main/java/com/hotels/shunting/yard/replicator/exec/app/ReplicationRunner.java

```diff
@@ -32,41 +35,63 @@
 import com.hotels.shunting.yard.replicator.exec.receiver.ReplicationMetaStoreEventListener;
 
 @Component
-class ReplicationRunner implements ApplicationRunner, ExitCodeGenerator {
+class ReplicationRunner implements ApplicationRunner, ExitCodeGenerator, Runnable {
   private static final Logger log = LoggerFactory.getLogger(ReplicationRunner.class);
 
   private static final Counter SUCCESS_COUNTER = Metrics.counter(MetricsConstant.RECEIVER_SUCCESSES);
   private static final Counter FAILURE_COUNTER = Metrics.counter(MetricsConstant.RECEIVER_FAILURES);
+  private static final Counter EMPTY_COUNTER = Metrics.counter(MetricsConstant.RECEIVER_EMPTY);
 
   private final ReplicationMetaStoreEventListener listener;
   private final MetaStoreEventReader eventReader;
+  private final TaskExecutor executor;
+  private boolean running = false;
 
   @Autowired
-  ReplicationRunner(MetaStoreEventReader eventReader, ReplicationMetaStoreEventListener listener) {
+  ReplicationRunner(
+      MetaStoreEventReader eventReader,
+      ReplicationMetaStoreEventListener listener,
+      TaskExecutor executor) {
     this.listener = listener;
     this.eventReader = eventReader;
+    this.executor = executor;
   }
 
   @Override
   public void run(ApplicationArguments args) {
-    while (eventReader.hasNext()) {
+    executor.execute(this);
+  }
+
+  @Override
+  public int getExitCode() {
+    return 0;
+  }
+
+  public void stop() {
```

Comment:
Will probably only ever be used by the unit test as Spring will handle stopping the thread if the application context is closed some other way. We will want to revisit this stop/start logic and should use a fancier executor that can stop the receiver from receiving new messages but wait until any messages in flight are completed before shutting down. For now this does the job and shouldn't be any worse than what we had before.

Reply:
Yeah, I agree it's not ideal but I'd rather not spend the time on it now. I think we need to sort out overall how this app runs and particularly how it shuts down so should sort all of that out at the same time and can improve this then. I'll make a note to create a ticket for this.